blog/_posts/2020-02-06-binding-redirects.markdown

```diff
@@ -0,0 +1,85 @@
+
+- [Enable `<AutoGenerateBindingRedirects>`](https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/how-to-enable-and-disable-automatic-binding-redirection) (this doesn't work for web projects - it doesn't handle `web.config`)
+  - Note: this isn't always perfect, it doesn't handle all transitive cases especially around multi-targeting and conditional references.
+- Build with Visual Studio and hope it has warnings and a click to fix (this *usually* works)
+  - Note: this isn't always perfect either, hence the "hope".
+- [**Manually edit your *.config file**](https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/redirect-assembly-versions#manually-editing-the-app-config-file).
+  - This is the surest way (and what Visual Studio is doing above), but also the most manual & fun on upgrades.
+
+Unfortunately, what that help manual editing doesn't mention is the ***assembly versions are not the NuGet versions***. For instance [System.Runtime.CompilerServices.Unsafe 4.7.0](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/4.7.0) on NuGet is _assembly_ version 4.0.6.0. The assembly version is what matters. The easiest way I use to figure this out on Windows is the [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer) ([Windows Store link](https://www.microsoft.com/en-us/p/nuget-package-explorer/9wzdncrdmdm3) - easiest install option) maintained by [Oren Novotny](https://github.com/onovotny). Either from within NPE's feed browser or from NuGet.org (there's a "Download package" option in the right sidebar), open the package. Select the framework you're loading (they should all match really) and double click the DLL. It'll have a section that says "Strong Name: Yes, version: `4.x.x.x`"
+
+For example, if you had libraries wanting various versions of `System.Numerics.Vectors` a most likely fix is to prefer the latest (as of writing this, [4.7.0 on NuGet](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/4.7.0) which is assembly version 4.0.6.0 from earlier) part of your config would look something like this:
+
+```xml
+<configuration>
+  <runtime>
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <dependentAssembly>
+        <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
```

Comment:
Maybe worth breaking down the attributes here? I don’t remember how it’s broken down in NPE but `publicKeyToken` is something people often get wrong. Also worth mentioning the `oldVersion` is a range but is generally always 0.0.0.0 to the latest to indicate _everything_ should be sent to `newVersion`.

Just random thoughts, everything else looks great!